### DISCUSSION TITLE
UniqueConstraint violation_error_message as error response in drf

### DISCUSSION THREAD

**Discussion by AGM-90** [encode/django-rest-framework#9352] [General]
Example Model
class UniqueTest(BaseModel):
    answer = models.ForeignKey(Answer, on_delete=models.CASCADE)
    subject = models.ForeignKey(Subject, on_delete=models.CASCADE)
    date = models.DateField()
   
    class Meta:
        constraints = [
            models.UniqueConstraint(
                fields=["answer", "subject","date"],
                condition=Q(deleted_at=None),
                name="uniq_if_not_del_pjt_site_sub_mil",
                violation_error_message="test error message"
            )
        ]
Sample error response for violation
{
    "non_field_errors": [
        "The fields answer, subject, date must make a unique set."
    ]
}
Expected error response
{
    "non_field_errors": [
       "test error message"
    ]
}

**Comment by s-aleshin** [comment]
Yes, I encountered this issue as well. When the serializer gathers constraints from the model, it doesn’t account for the fact that a constraint might include a custom error message. Additionally, the `UniqueTogetherValidator` is created with a default message. https://github.com/encode/django-rest-framework/blob/master/rest_framework/serializers.py#L1595-L1621

**Comment by igdimer** [comment]
Hi! I face to the same issue.

**Comment by soldatov-ss** [comment]
guys, any news? short workarounds?

**Reply by s-aleshin** [reply]
I created a custom validator by subclassing `UniqueTogetherValidator` and overriding its `__call__` method. This allows me to catch the default error and raise my own `ValidationError` with a custom message and error code.
```python
class CustomUniqueTogetherValidator(UniqueTogetherValidator):
    def __call__(self, attrs, serializer):
        self.enforce_required_fields(attrs, serializer)
        try:
            super().__call__(attrs, serializer)
        except serializers.ValidationError as exc:
            [msg] = exc.detail
            raise serializers.ValidationError(
                str(msg), code="custom_unique_constraint"
            ) from exc


class MySerializer(serializers.ModelSerializer):
    class Meta:
        model = MyModel
        fields = "__all__"
        validators = [
            CustomUniqueTogetherValidator(
                queryset=MyModel.objects.filter(...).all(),
                fields=["field1", "field2"],
                message="My custom message."
            ),
        ]
```
⚠️ When you override the validators attribute in your serializer’s Meta, it replaces all automatically generated validators based on model constraints (https://github.com/encode/django-rest-framework/blob/master/rest_framework/serializers.py#L1561-L1563). DRF will no longer auto-generate them — so you must explicitly define any additional validators you need (e.g., for other UniqueConstraints or CheckConstraints).

If an issue is opened based on this discussion, I’d be happy to submit a PR with a fix.

**Comment by efefre** [comment]
I have a very similar situation to the one described, with the difference that I have a `UniqueConstraint` with a single field in `fields` and a `condition`. In this case, unfortunately, `violation_error_message` is ignored.

Unfortunately, in https://github.com/encode/django-rest-framework/pull/9766, this case was not taken into account.
